src/Evolution/Systems/Ccz4/TimeDerivative.cpp

```diff
@@ -0,0 +1,625 @@
+    const tnsr::ii<DataVector, Dim>& conformal_spatial_metric,
+    const Scalar<DataVector>& ln_lapse, const tnsr::I<DataVector, Dim>& shift,
+    const Scalar<DataVector>& ln_conformal_factor,
+    const tnsr::ii<DataVector, Dim>& a_tilde,
+    const Scalar<DataVector>& trace_extrinsic_curvature,
+    const Scalar<DataVector>& theta, const tnsr::I<DataVector, Dim>& gamma_hat,
+    const tnsr::I<DataVector, Dim>& b, const tnsr::i<DataVector, Dim>& field_a,
+    const tnsr::iJ<DataVector, Dim>& field_b,
+    const tnsr::ijj<DataVector, Dim>& field_d,
+    const tnsr::i<DataVector, Dim>& field_p,
+    // spatial derivatives of evolved variables
+    const tnsr::ijj<DataVector, Dim>& d_a_tilde,
+    const tnsr::i<DataVector, Dim>& d_trace_extrinsic_curvature,
+    const tnsr::i<DataVector, Dim>& d_theta,
+    const tnsr::iJ<DataVector, Dim>& d_gamma_hat,
+    const tnsr::iJ<DataVector, Dim>& d_b,
+    const tnsr::ij<DataVector, Dim>& d_field_a,
+    const tnsr::ijK<DataVector, Dim>& d_field_b,
+    const tnsr::ijkk<DataVector, Dim>& d_field_d,
+    const tnsr::ij<DataVector, Dim>& d_field_p) {
+  ASSERT(
+      evolve_shift == EvolveShift::True or evolve_shift == EvolveShift::False,
+      "Unknown Ccz4::EvolveShift.");
+
+  constexpr double one_third = 1.0 / 3.0;
+
+  // quantities we need for computing eq 12 - 27
+
+  determinant_and_inverse(det_conformal_spatial_metric,
+                          inv_conformal_spatial_metric,
+                          conformal_spatial_metric);
+
+  const size_t num_points = get_size(get(ln_conformal_factor));
+  for (size_t i = 0; i < num_points; i++) {
+    get(*conformal_factor_squared)[i] = exp(2.0 * get(ln_conformal_factor)[i]);
+  }
+
+  ::TensorExpressions::evaluate<ti_I, ti_J>(
+      inv_spatial_metric, (*conformal_factor_squared)() *
+                              (*inv_conformal_spatial_metric)(ti_I, ti_J));
+
+  ::TensorExpressions::evaluate<ti_I, ti_J>(
+      inv_a_tilde, a_tilde(ti_k, ti_l) *
+                       (*inv_conformal_spatial_metric)(ti_I, ti_K) *
+                       (*inv_conformal_spatial_metric)(ti_J, ti_L));
+
+  for (size_t i = 0; i < num_points; i++) {
+    get(*lapse)[i] = exp(get(ln_lapse)[i]);
+  }
+
+  if (slicing_condition_type == SlicingConditionType::Harmonic) {
+    get(*slicing_condition) = 1.0;
+    get(*d_slicing_condition) = 0.0;
+    get(*lapse_times_slicing_condition) = get(*lapse);
+  } else if (slicing_condition_type == SlicingConditionType::Log) {
+    ASSERT(min(get(*lapse)) > 0.0,
+           "The lapse must be positive when using "
+           "Ccz4::SlicingConditionType::Harmonic.");
+
+    // g(\alpha) == 2 / \alpha
+    get(*slicing_condition) = 2.0 / get(*lapse);
+    // g'(\alpha)  == -2 / \alpha^2 == -0.5 * g(\alpha)^2
+    get(*d_slicing_condition) = -0.5 * square(get(*slicing_condition));
+    // \alpha g(\alpha)  == 2
+    get(*lapse_times_slicing_condition) = 2.0;
+  } else {
+    ERROR("Unknown Ccz4::SlicingConditionType");
```

Comment:
Commenting here so the placement is clear. This is an extremely unlikely branch to hit. I would actually consider decorating the first if as `UNLIKELY` and the second as `LIKELY` so that the critical path in binaries is the speculatively evaluated one. The reason it's not crucial to have a test is because it's very unlikely someone will produce an `enum class` value that isn't one of the known ones. `enum classes` don't allow implicit conversions from their underlying integral type. You need to do a `static_cast` and at that point the person either knows what they're doing or has earned learning to know why they shouldn't do what they're doing :)

The "NOLINT-equivalent" statement is make this like:
```cpp
 ERROR("Unknown Ccz4::SlicingConditionType");  // LCOV_EXCL_LINE
```
This way code coverage will say "Ah, we don't expect this line to be covered. All good." Hitting this error means the person has tried really hard to write extremely bad C++ (ignored multiple compiler errors). This `LCOV_EXCL_LINE` is the only change being requested.